### ISSUE TITLE
BG 356: Correct TrackDefinition?

### ISSUE THREAD

**Issue by BerndHekele** [openETCS/modeling#731]
Passing BG 365 we get packet 41 (Level Transition Order to Level NTC).
This looks a bit misplaced. 
The EVC reacts and goes immediatly in trip (due to m missing MA.

@BaseliyosJacob : What is the expected (and recorded) Behaviour of the train?
Shall we change the track to get rid of additional NTC change requests?

**Comment by janWelte** [comment]
BG 365 should not transmit any P41.
A P41 is first placed in BG 440 to switch back.

But I've not been able to confirm the P 41 in BG 365 or any M03 radio message around this point.

**Comment by BaseliyosJacob** [comment]
Dear Jan,

this are my information about the content of the BG 356 be care: not BG 365 - i assume that bernd has switched by accident the 6 and 5.

**Comment by BaseliyosJacob** [comment]
![image](https://cloud.githubusercontent.com/assets/3975810/9885818/73b4e93c-5be7-11e5-83d9-4ff476dc33f1.png)

**Comment by BaseliyosJacob** [comment]
@BerndHekele The Ack is sent twice because

P41 
 • first time triggered by the radio message
 • second time triggered by the balise message in BG 356
 
the trigger by the balise trigger is just repeating the level transition for the case that the transition has not been triggeres by the RBC message or for level 1. But should be not considered since the first from the RBC has been accepted. This should be from my point of view handled in the Level and Mode Management. See # https://github.com/openETCS/modeling/issues/693

**Comment by BerndHekele** [comment]
Hi Baseliyos,thanks for the correction: we talk about balise 356.
This information is not repeated by radio and it is requesting to switch to NTC.
There is no information in the P41 related to level 2.
From my point of view this is far to early to enter the station. 
If we stick to this track-layout we either need to ignore the level change to NTC (DMI-Ack = false).
In this case we need to change the behaviour of the emergency brake to continue the trip (not the mode).

Can you explain what happened in the real trip at this point.

**Comment by janWelte** [comment]
As I've said before, this is the P41 to enter Amsterdam as Q_DIr is "nominal" and your are passing it in reverse.

Therefore the packet is set to invalid by the function to validate data direction. If i see it right your bypass is just checking for P41 and therefore does not see that this package is not valid for the train.

**Comment by janWelte** [comment]
Behaviour in the JRU log is just as it should be at a linked balise
1. position report
2. repositioning of targets

**Comment by BerndHekele** [comment]
Thanks Jan for clarification.
Indeed we have an error in the workaround for handling the p41 valid indicator.
However, the scenario is relevant for completing the trip to the station Utrecht.
Therefore I report current status of this transition with a target to prepare the eom.

For Switching to NTC Levels needs a L!_MA and target speed. Both information is not visible in the BG-packets.

@BaseliyosJacob 
Where does the train  get this information from (in the real scenario).

**Comment by BerndHekele** [comment]
@janWelte , @mairamou 
Update after putting in some correction in the workaround. (the correction is tested).
We are passing the Balise in nominal direction:
![image](https://cloud.githubusercontent.com/assets/3288918/9908847/d26d6432-5c95-11e5-8262-3dcbc96f1551.png)

This implies we need to correct the track definition for this BG.

Can you please confirm.

kr
Bernd

**Comment by janWelte** [comment]
This is a track implementation error in SCADE. 
The JRU clearly shows

 Time | Message | ... | N_PIG | N_TOTAL | M_DUP | M_Count | NID_C | country | p | NID_BG | Q_Link | P  
  --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | ---  | ---  
7078369	| > M6 <  TELEGRAM FROM BALISE(6)	| ... |N_PIG(3bits) = I am the 2nd(1)	|N_TOTAL(3bits) = 2 balises in the group(1)	| M_DUP(2bits) = This balise is a duplicate of the previous balise(2) |	M_MCOUNT(8bits) = The telegram fits with all telegrams of the same balise group(255) | NID_C(10bits) = 426	| Netherlands	| various projects(426)	|NID_BG(14bits) = 356(356)	|Q_LINK(1bits) = Linked(1)	|> P41 <
7078370	|> M6 <  TELEGRAM FROM BALISE(6)| ... |N_PIG(3bits) = I am the 1st(0) | N_TOTAL(3bits) = 2 balises in the group(1)	| M_DUP(2bits) = This balise is a duplicate of the next balise(1)	| M_MCOUNT(8bits) = The telegram fits with all telegrams of the same balise group(255)	| NID_C(10bits) = 426	|  Netherlands	| various projects(426)	|NID_BG(14bits) = 356(356)	|Q_LINK(1bits) = Linked(1)	|> P41 <

But I can not confirm your that the track is implemented wrong, as the balises are placed correct in the BG. (First Balise 1 and second Balise 0)

**Comment by janWelte** [comment]
Bernd is seams that you are correct, there is a calculation error in the Balise interdistance calculation. 
This function always sends the BG as nominal even if the balise is correctly implemented as reverse.
@mairamou please confirm this an correct it. 
If the first balise is the B1 it should be send in front of the BG position not the other way around.

**Comment by mairamou** [comment]
ok 

I will look at it. 

On 2015-09-16 19:24, Jan Welte wrote: 

> Bernd is seams that you are correct, there is a calculation error in the Balise interdistance calculation. 
> This function always sends the BG as nominal even if the balise is correctly implemented as reverse.
> @mairamou [1] please confirm this an correct it. 
> If the first balise is the B1 it should be send in front of the BG position not the other way around. 
> 
> --
> Reply to this email directly or view it on GitHub [2].

-- 
Mairamou Haman Adji
Junior Software Developer
Research Fellow
---------------------------
LEA RAILERGY
 

Links:
------
[1] https://github.com/mairamou
[2]
https://github.com/openETCS/modeling/issues/731#issuecomment-140832748

**Comment by BerndHekele** [comment]
We are urgently waiting for the fix.
When will it be available?

**Comment by mairamou** [comment]
Hello, 

sorry, I was (very) sick, sorry for the delay. 

I am working on the solution and expect it for later today. 

Mairamou 

On 2015-09-21 09:59, Bernd Hekele wrote: 

> We are urgently waiting for the fix.
> When will it be available? 
> 
> --
> Reply to this email directly or view it on GitHub [1].

-- 
Mairamou Haman Adji
Junior Software Developer
Research Fellow
---------------------------
LEA RAILERGY
 

Links:
------
[1]
https://github.com/openETCS/modeling/issues/731#issuecomment-141902701

**Comment by mairamou** [comment]
I have provided a solution that I have tested at BG 354 for nominal and BG261 for reverse.
I hope it is working now. (the library operator was not my design). 
It seems that there were was a some double negation (using a factor of -1 where this was already defined by the balise locations inside the BG effectively cancelling the inverse positions). 

I am however at the moment not sure that my solution works for BGs with more than two balises, but this is less urgent for now. (waiting for the repository to clone because I had some consistency issues)
